precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+		let input = InputReader::new(input)?;
+
+		match input.selector() {
+			[0x7c, 0x80, 0xaa, 0x9f] => Self::total_supply(input),
+			[0x70, 0xa0, 0x82, 0x31] => Self::balance_of(input),
+			[0xdd, 0x62, 0xed, 0x3e] => Self::allowance(input),
+			[0x09, 0x5e, 0xa7, 0xb3] => Self::approve(input, context),
+			[0xa9, 0x05, 0x9c, 0xbb] => Self::transfer(input, context, target_gas),
+			[0x0c, 0x41, 0xb0, 0x33] => Self::transfer_from(input, context, target_gas),
+			_ => Err(error("unknown selector")),
+		}
+	}
+}
+
+impl<Runtime, Instance> Erc20BalancesPrecompile<Runtime, Instance>
+where
+	Runtime: pallet_balances::Config<Instance> + pallet_evm::Config,
+	Runtime::AccountId: From<H160>,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::Call: From<pallet_balances::Call<Runtime, Instance>>,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	Instance: InstanceToPrefix + 'static,
+	U256: From<BalanceOf<Runtime, Instance>> + TryInto<BalanceOf<Runtime, Instance>>,
+{
+	fn total_supply(input: InputReader) -> EvmResult<PrecompileOutput> {
+		// Parse input.
+		input.expect_arguments(0)?;
+
+		// Fetch info.
+		let amount = pallet_balances::Pallet::<Runtime, Instance>::total_issuance();
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: RuntimeHelper::<Runtime>::db_read_gas_cost(),
+			output: OutputBuilder::new().write_u256(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn balance_of(mut input: InputReader) -> EvmResult<PrecompileOutput> {
+		// Read input.
+		input.expect_arguments(1)?;
+
+		let address = input.read_address()?;
+
+		// Fetch info.
+		let amount = pallet_balances::Pallet::<Runtime, Instance>::usable_balance(&address.into());
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: RuntimeHelper::<Runtime>::db_read_gas_cost(),
+			output: OutputBuilder::new().write_u256(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn allowance(mut input: InputReader) -> EvmResult<PrecompileOutput> {
+		// Read input.
+		input.expect_arguments(2)?;
+
+		let owner_id: Runtime::AccountId = input.read_address()?.into();
+		let spender_id: Runtime::AccountId = input.read_address()?.into();
+
+		// Fetch info.
+		let amount =
+			ApprovesStorage::<Runtime, Instance>::get(owner_id, spender_id).unwrap_or_default();
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: RuntimeHelper::<Runtime>::db_read_gas_cost(),
+			output: OutputBuilder::new().write_u256(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn approve(mut input: InputReader, context: &Context) -> EvmResult<PrecompileOutput> {
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let spender = input.read_address()?;
+		let amount = Self::u256_to_amount(input.read_u256()?)?;
+
+		let spender_id: Runtime::AccountId = spender.into();
+		let caller_id: Runtime::AccountId = context.caller.into();
+
```

Comment:
Instead `db_read_gas_cost`, do `db_write_gas_cost`. Check also if `OutOfGas`, return `Error`.

Include this in the tests.